Update invokedynamic and invokehandle bytecode handlers

nbhuiyan opened the pull request:
This changeset updates the handling of `invokedynamic` and `invokehandle`
bytecodes in preparation for migrating to the OpenJDK MethodHandle
implementation. The changes are not enabled by default, and are guarded
in the macro `J9VM_OPT_OPENJDK_METHODHANDLE`.

This implementation can handle both resolved and unresolved
`invokedynamic` and `invokehandle` bytecodes.

Signed-off-by: Nazim Bhuiyan <nubhuiyan@ibm.com>

nbhuiyan commented:
@liqunl @andrewcraik Requesting review.

andrewcraik commented:
@liqunl will let you make a first pass on this one due to your expertise in the area and then I'll do a review - thanks!

andrewcraik commented:
@liqunl friendly reminder on the review request

liqunl commented:
Will look today

liqunl commented on runtime/compiler/compile/J9SymbolReferenceTable.cpp:
```diff
@@ -404,18 +408,37 @@ J9::SymbolReferenceTable::findOrCreateHandleMethodSymbol(TR::ResolvedMethodSymbo
    TR_ResolvedMethod *owningMethod = owningMethodSymbol->getResolvedMethod();
+#if defined(J9VM_OPT_OPENJDK_METHODHANDLE)
+   J9InvokeCacheEntry *invokeCache = (J9InvokeCacheEntry *) owningMethod->callSiteTableEntryAddress(callSiteIndex);
+   if (!invokeCache)
```
`invokeCache` is unlikely to be null based on the way it's calculated. The exception is unnecessary. If it's null, there will be a functional issue and we'll crash trying to access its member variable.

liqunl commented on runtime/compiler/compile/J9SymbolReferenceTable.cpp:
```diff
@@ -453,12 +476,27 @@ J9::SymbolReferenceTable::findOrCreateCallSiteTableEntrySymbol(TR::ResolvedMetho
    TR_ResolvedMethod *owningMethod = owningMethodSymbol->getResolvedMethod();
+#if defined(J9VM_OPT_OPENJDK_METHODHANDLE)
+   J9InvokeCacheEntry *invokeCache = (J9InvokeCacheEntry *) owningMethod->methodTypeTableEntryAddress(cpIndex);
+   if (!invokeCache)
```
Same here.

liqunl commented on runtime/compiler/env/j9method.cpp:
```diff
@@ -6757,6 +6757,39 @@ TR_ResolvedJ9Method::getResolvedDynamicMethod(TR::Compilation * comp, I_32 callS
    TR_ResolvedMethod *result = NULL;
 
+#if defined(J9VM_OPT_OPENJDK_METHODHANDLE)
+   TR_ResolvedMethod * owningMethod = comp->getCurrentMethod();
```
`owningMethod` should be `this`. We're looking at invokedynamic bytecode of `this` resolved method.

liqunl commented on runtime/compiler/env/j9method.cpp:
```diff
@@ -6757,6 +6757,39 @@ TR_ResolvedJ9Method::getResolvedDynamicMethod(TR::Compilation * comp, I_32 callS
+#if defined(J9VM_OPT_OPENJDK_METHODHANDLE)
+   TR_ResolvedMethod * owningMethod = comp->getCurrentMethod();
+   J9InvokeCacheEntry *invokeCache = (J9InvokeCacheEntry *) owningMethod->callSiteTableEntryAddress(callSiteIndex);
+   if (!invokeCache)
```
No need to fail compilation here.

liqunl commented on runtime/compiler/env/j9method.cpp:
```diff
@@ -6757,6 +6757,39 @@ TR_ResolvedJ9Method::getResolvedDynamicMethod(TR::Compilation * comp, I_32 callS
+   if (!invokeCache)
+      comp->failCompilation<TR::CompilationException>("unable to retrieve side table entry for invokeDynamic");
+      {
+      TR::VMAccessCriticalSection getResolvedDynamicMethod(fej9());
```
Reading constant pool entries does require a VM access. I think the VM access including the original one can be removed.

liqunl commented on runtime/compiler/env/j9method.cpp:
```diff
@@ -6757,6 +6757,39 @@ TR_ResolvedJ9Method::getResolvedDynamicMethod(TR::Compilation * comp, I_32 callS
+      J9Class    *ramClass = constantPoolHdr();
+      J9ROMClass *romClass = romClassPtr();
+
+      if (ramClass->callSites[callSiteIndex] == NULL)
```
We should use `isUnresolvedCallSiteTableEntry` to see if entry is unresolved.

liqunl commented on runtime/compiler/env/j9method.cpp:
```diff
@@ -6757,6 +6757,39 @@ TR_ResolvedJ9Method::getResolvedDynamicMethod(TR::Compilation * comp, I_32 callS
+         J9ROMNameAndSignature *nameAndSig   = NNSRP_GET(namesAndSigs[callSiteIndex], J9ROMNameAndSignature*);
+         J9UTF8                *signature    = J9ROMNAMEANDSIGNATURE_SIGNATURE(nameAndSig);
+
+         TR_OpaqueMethodBlock *dummyInvoke = _fe->getMethodFromName("java/lang/invoke/MethodHandle", "linkToStatic", "([Ljava/lang/Object;)Ljava/lang/Object;");
```
I think the dummy invoke is the only thing different from current implementation, other code can be shared

liqunl commented on runtime/compiler/env/j9method.cpp:
```diff
@@ -6757,6 +6757,39 @@ TR_ResolvedJ9Method::getResolvedDynamicMethod(TR::Compilation * comp, I_32 callS
+         }
+      else
+         {
+         TR_OpaqueMethodBlock * targetJ9MethodBlock = fej9()->targetMethodFromMemberName((uintptr_t) invokeCache->target);
```
VM access is needed to call `targetMethodFromMemberName` with object reference.

liqunl commented on runtime/compiler/env/j9method.cpp:
```diff
@@ -6802,6 +6835,39 @@ TR_ResolvedJ9Method::getResolvedHandleMethod(TR::Compilation * comp, I_32 cpInde
 #else
    TR_ResolvedMethod *result = NULL;
 
+#if defined(J9VM_OPT_OPENJDK_METHODHANDLE)
```
Comments for `getResolvedDynamicMethod` apply here

liqunl commented:
Do you forget to include the changes to `methodTypeTableEntryAddress` and `callSiteTableEntryAddress`? The entry address calculation should be changed because the two tables have been changed to contain `J9InvokeCacheEntry` (equivalent to two objects) per entry instead of a object.

liqunl commented on runtime/compiler/optimizer/InterpreterEmulator.cpp:
```diff
@@ -627,7 +662,53 @@ InterpreterEmulator::visitInvokedynamic()
+   TR_ResolvedMethod * owningMethod = _methodSymbol->getResolvedMethod();
+
+   J9InvokeCacheEntry *invokeCache = (J9InvokeCacheEntry *) owningMethod->methodTypeTableEntryAddress(cpIndex);
+   if (!invokeCache) return; // unresolved
```
Same here, should use `isUnresolvedMethodTypeTableEntry`

liqunl commented on runtime/compiler/optimizer/InterpreterEmulator.cpp:
```diff
@@ -604,6 +607,38 @@ InterpreterEmulator::visitInvokedynamic()
+   TR_ResolvedMethod * owningMethod = _methodSymbol->getResolvedMethod();
+
+   J9InvokeCacheEntry *invokeCache = (J9InvokeCacheEntry *) owningMethod->callSiteTableEntryAddress(callSiteIndex);
+   if (!invokeCache) return; // unresolved
```
`invokeCache` will never be null, it's just a pointer into the side table. We should use `isUnresolvedCallSiteTableEntry`

liqunl commented on runtime/compiler/ilgen/Walker.cpp:
```diff
@@ -3352,6 +3379,48 @@ TR_J9ByteCodeIlGenerator::genInvokeHandle(TR::SymbolReference *invokeExactSymRef
+      // additional parameters.
+      TR::SymbolReference *memberNameSymRef = symRefTab()->findOrCreateCallSiteTableEntrySymbol(_methodSymbol, callSiteIndex, true);
+      TR::Node * memberNameNode = loadSymbol(TR::aload, memberNameSymRef);
+      appendixNode->setIsNull(true);
```
Setting null flag on a node means that the node produce null value at runtime, which is wrong.

liqunl commented on runtime/compiler/ilgen/Walker.cpp:
```diff
@@ -3352,6 +3379,48 @@ TR_J9ByteCodeIlGenerator::genInvokeHandle(TR::SymbolReference *invokeExactSymRef
+      // additional parameters.
+      TR::SymbolReference *memberNameSymRef = symRefTab()->findOrCreateMethodTypeTableEntrySymbol(_methodSymbol, cpIndex, true);
+      TR::Node * memberNameNode = loadSymbol(TR::aload, memberNameSymRef);
+      appendixNode->setIsNull(true);
```
Same here.

liqunl commented on runtime/compiler/ilgen/Walker.cpp:
```diff
@@ -3272,6 +3272,21 @@ TR_J9ByteCodeIlGenerator::genInvokeDynamic(int32_t callSiteIndex)
+
+   TR::Node* callNode = genInvokeDirect(targetMethodSymRef);
+
+   _invokeDynamicCalls->set(_bcIndex);
```
The bit vector is used in the openj9 implementation to transform the call later. The new implementation doesn't need it.

nbhuiyan commented:
@liqunl the changes to `methodTypeTableEntryAddress` and `callSiteTableEntryAddress` that I had in my working branch were not mine originally, so I thought they would be part of a different PR -- either yours or part of the PRs introducing the VM changes. Since that's not the case, I will include them in this PR.

liqunl commented:
> @liqunl the changes to `methodTypeTableEntryAddress` and `callSiteTableEntryAddress` that I had in my working branch were not mine originally, so I thought they would be part of a different PR -- either yours or part of the PRs introducing the VM changes. Since that's not the case, I will include them in this PR.

Yes, please include them in this  PR

andrewcraik commented:
You can add a co-authored entry on the commit and PR to make it clear if @liqunl helped write some of that code. I agree with keep the commits logical and self-contained.

nbhuiyan commented:
Ready for another round of review.

andrewcraik commented:
@liqunl could you do another pass now that the PR has been updated - thanks @nbhuiyan !

liqunl commented on runtime/compiler/env/j9method.cpp:
```diff
@@ -6766,30 +6775,52 @@ TR_ResolvedJ9Method::getResolvedDynamicMethod(TR::Compilation * comp, I_32 callS
+   J9ROMNameAndSignature *nameAndSig   = NNSRP_GET(namesAndSigs[callSiteIndex], J9ROMNameAndSignature*);
+   J9UTF8                *signature    = J9ROMNAMEANDSIGNATURE_SIGNATURE(nameAndSig);
+
+#if defined(J9VM_OPT_OPENJDK_METHODHANDLE)
+   TR_OpaqueMethodBlock *dummyInvoke = _fe->getMethodFromName("java/lang/invoke/MethodHandle", "linkToStatic", "([Ljava/lang/Object;)Ljava/lang/Object;");
+   char * linkToStaticSignature = (char *) comp->trMemory()->allocateMemory((J9UTF8_LENGTH(signature)+40)*sizeof(char), heapAlloc);
+   char * romMethodSignature = (char *) comp->trMemory()->allocateMemory((J9UTF8_LENGTH(signature))*sizeof(char), heapAlloc);
+   strcpy(romMethodSignature, utf8Data(signature));
+   char * sigTokenStart = strtok(romMethodSignature, ")");
+   char * sigTokenEnd = strtok(NULL, ")");
+   sprintf(linkToStaticSignature, "%sLjava/lang/Object;Ljava/lang/Object;)%s", sigTokenStart, sigTokenEnd);
+   result = _fe->createResolvedMethodWithSignature(comp->trMemory(), dummyInvoke, NULL, linkToStaticSignature, strlen(linkToStaticSignature), this);
```
You don't need to allocate new memory to store the signature. Like the original code, you can do
```
TR_OpaqueMethodBlock *dummyInvoke = _fe->getMethodFromName("java/lang/invoke/MethodHandle", "linkToStatic", "([Ljava/lang/Object;)Ljava/lang/Object;");
result = _fe->createResolvedMethodWithSignature(comp->trMemory(), dummyInvoke, NULL, utf8Data(signature), J9UTF8_LENGTH(signature), this);
```

nbhuiyan replied:
When the call is unresolved, we create a dummy invoke to `linkToStatic`. `linkToStatic` is signature polymorphic and therefore can have any number of args for the call, and return any type. We also need to provide the `appendix` and `target` objects as the last two args. `utf8Data(signature) will just get us the call site signature without accounting for the 2 objects we are providing as args.` Therefore, we take the ROM method signature, and split it where we need to insert the 2 additional parameters in the signature. `strtok` modifies the strings, which is why we make a writable copy of it to work on. I will add some comments explaining the rationale behind this and also reduce the duplication with `invokeHandle`.

liqunl commented on runtime/compiler/env/j9method.cpp:
```diff
@@ -6811,21 +6841,45 @@ TR_ResolvedJ9Method::getResolvedHandleMethod(TR::Compilation * comp, I_32 cpInde
+   J9ROMNameAndSignature *nameAndSig = J9ROMMETHODREF_NAMEANDSIGNATURE(romMethodRef);
+
+#if defined(J9VM_OPT_OPENJDK_METHODHANDLE)
+   J9UTF8                *signature    = J9ROMNAMEANDSIGNATURE_SIGNATURE(nameAndSig);
+   J9InvokeCacheEntry *invokeCache = (J9InvokeCacheEntry *) methodTypeTableEntryAddress(cpIndex);
+
+   if (!isUnresolvedMethodTypeTableEntry(cpIndex))
       {
-      TR::VMAccessCriticalSection getResolvedHandleMethod(fej9());
+      TR_OpaqueMethodBlock * targetJ9MethodBlock = NULL;
+         {
+         VMAccessCriticalSection getResolvedHandleMethod(fej9());
+         targetJ9MethodBlock = fej9()->targetMethodFromMemberName((uintptr_t) invokeCache->target);
+         }
+      result = fej9()->createResolvedMethod(comp->trMemory(), targetJ9MethodBlock, this);
+      return result;
+      }
+   TR_OpaqueMethodBlock *dummyInvoke = _fe->getMethodFromName("java/lang/invoke/MethodHandle", "linkToStatic", "([Ljava/lang/Object;)Ljava/lang/Object;");
+   char * linkToStaticSignature = (char *) comp->trMemory()->allocateMemory((J9UTF8_LENGTH(signature)+40)*sizeof(char), heapAlloc);
+   char * romMethodSignature = (char *) comp->trMemory()->allocateMemory((J9UTF8_LENGTH(signature))*sizeof(char), heapAlloc);
+   strcpy(romMethodSignature, utf8Data(signature));
+   char * sigTokenStart = strtok(romMethodSignature, ")");
+   char * sigTokenEnd = strtok(NULL, ")");
+   sprintf(linkToStaticSignature, "%sLjava/lang/Object;Ljava/lang/Object;)%s", sigTokenStart, sigTokenEnd);
+   result = _fe->createResolvedMethodWithSignature(comp->trMemory(), dummyInvoke, NULL, linkToStaticSignature, strlen(linkToStaticSignature), this);
```
Same problem here

nbhuiyan replied:
See my response to the other comment.

liqunl commented on runtime/compiler/env/j9method.cpp:
```diff
@@ -6766,30 +6775,52 @@ TR_ResolvedJ9Method::getResolvedDynamicMethod(TR::Compilation * comp, I_32 callS
-      J9Class    *ramClass = constantPoolHdr();
-      J9ROMClass *romClass = romClassPtr();
+#if defined(J9VM_OPT_OPENJDK_METHODHANDLE)
+   J9InvokeCacheEntry *invokeCache = (J9InvokeCacheEntry *) callSiteTableEntryAddress(callSiteIndex);
```
There are a lot of duplication of this code. Can we add a few APIs to get address of member name/appendix and hide the implementation detail of the side table? Something like
```
memberNameAddressFromInvokeDynamicSideTable(int32_t callSiteIndex)
appendixAddressFromInvokeDynamicSideTable(int32_t callSiteIndex)
memberNameAddressFromInvokeHandleSideTable(int32_t cpIndex)
appendixAddressFromInvokeHandleSideTable(int32_t cpIndex)
```

nbhuiyan replied:
Done.

nbhuiyan commented:
@liqunl I have addressed your comments and tested the implementation of both `invokedynamic` and `invokehandle` bytecode handlers locally with the latest changes.

andrewcraik commented:
@liqunl have your concerns been addressed?

liqunl commented on runtime/compiler/env/j9method.cpp:
```diff
@@ -6902,6 +6943,36 @@ TR_ResolvedJ9Method::fieldsAreSame(I_32 cpIndex1, TR_ResolvedMethod * m2, I_32 c
+TR_ResolvedJ9Method::memberNameAddressFromInvokeDynamicSideTable(int32_t callSiteIndex)
+   {
+   J9InvokeCacheEntry *invokeCache = (J9InvokeCacheEntry *) callSiteTableEntryAddress(callSiteIndex);
+   return (uintptr_t) invokeCache->target;
```
This function should return address of the object reference in the table. `invokeCache->target` is the actual object reference, we should return  `&invokeCache->target`. Like `methodTypeTableEntryAddress` and `callSiteTableEntryAddress`, the return type `void *`

liqunl commented on runtime/compiler/env/j9method.cpp:
```diff
@@ -6902,6 +6943,36 @@ TR_ResolvedJ9Method::fieldsAreSame(I_32 cpIndex1, TR_ResolvedMethod * m2, I_32 c
+TR_ResolvedJ9Method::appendixAddressFromInvokeDynamicSideTable(int32_t callSiteIndex)
+   {
+   J9InvokeCacheEntry *invokeCache = (J9InvokeCacheEntry *) callSiteTableEntryAddress(callSiteIndex);
+   return (uintptr_t) invokeCache->appendix;
```
Same here

liqunl commented on runtime/compiler/env/j9method.cpp:
```diff
@@ -6902,6 +6943,36 @@ TR_ResolvedJ9Method::fieldsAreSame(I_32 cpIndex1, TR_ResolvedMethod * m2, I_32 c
+TR_ResolvedJ9Method::memberNameAddressFromInvokeHandleSideTable(int32_t cpIndex)
+   {
+   J9InvokeCacheEntry *invokeCache = (J9InvokeCacheEntry *) methodTypeTableEntryAddress(cpIndex);
+   return (uintptr_t) invokeCache->target;
```
Same here

liqunl commented on runtime/compiler/env/j9method.cpp:
```diff
@@ -6902,6 +6943,36 @@ TR_ResolvedJ9Method::fieldsAreSame(I_32 cpIndex1, TR_ResolvedMethod * m2, I_32 c
+TR_ResolvedJ9Method::appendixAddressFromInvokeHandleSideTable(int32_t cpIndex)
+   {
+   J9InvokeCacheEntry *invokeCache = (J9InvokeCacheEntry *) methodTypeTableEntryAddress(cpIndex);
+   return (uintptr_t) invokeCache->appendix;
```
Same here

liqunl commented on runtime/compiler/optimizer/InterpreterEmulator.cpp:
```diff
@@ -603,13 +606,40 @@ InterpreterEmulator::visitInvokedynamic()
+   TR_ResolvedJ9Method * owningMethod = static_cast<TR_ResolvedJ9Method*>(_methodSymbol->getResolvedMethod());
+
+#if defined(J9VM_OPT_OPENJDK_METHODHANDLE)
+   if (!knot && owningMethod->isUnresolvedCallSiteTableEntry(callSiteIndex)) return;
```
Should this be `||`? Also, I don't think we need `knot` to inline the adapter method. Adapter method is static and is known if the callsite is resolved.

liqunl commented on runtime/compiler/optimizer/InterpreterEmulator.cpp:
```diff
@@ -603,13 +606,40 @@ InterpreterEmulator::visitInvokedynamic()
+
+   // add appendix object to knot and push to stack
+   uintptr_t appendixAddress = owningMethod->appendixAddressFromInvokeDynamicSideTable(callSiteIndex);
+   push(new (trStackMemory()) KnownObjOperand(knot->getOrCreateIndexAt(&appendixAddress)));
```
If `appendixAddressFromInvokeDynamicSideTable` returns the address instead of the appendix object reference, we don't need to take the address of it.

If `knot` is null, just push unknown operand.

liqunl commented on runtime/compiler/optimizer/InterpreterEmulator.cpp:
```diff
@@ -603,13 +606,40 @@ InterpreterEmulator::visitInvokedynamic()
+   TR_OpaqueMethodBlock* targetMethodObj = 0;
+      {
+      TR::VMAccessCriticalSection vmAccess(fej9);
+      targetMethodObj = fej9->targetMethodFromMemberName(owningMethod->memberNameAddressFromInvokeDynamicSideTable(callSiteIndex));
```
If `memberNameAddressFromInvokeDynamicSideTable` returns an address, we need to dereference the address to get the MemberName. Or you can add a function `targetMethodFromMemberNameAt` to take address of MemberName, and return you the target method. Inside `targetMethodFromMemberNameAt`, you need to acquire vm access

liqunl commented on runtime/compiler/optimizer/InterpreterEmulator.cpp:
```diff
@@ -621,13 +651,59 @@ InterpreterEmulator::visitInvokedynamic()
+
+#if defined(J9VM_OPT_OPENJDK_METHODHANDLE)
+void
+InterpreterEmulator::visitInvokehandle()
```
Comments for `visitInvokedynamic` apply to `visitInvokehandle`

liqunl commented on runtime/compiler/optimizer/InterpreterEmulator.cpp:
```diff
@@ -621,13 +651,59 @@ InterpreterEmulator::visitInvokedynamic()
+   // add appendix object to knot and push to stack
+   TR::KnownObjectTable *knot = comp()->getOrCreateKnownObjectTable();
+   uintptr_t appendixAddress = owningMethod->appendixAddressFromInvokeHandleSideTable(cpIndex);
+   if (knot) push(new (trStackMemory()) KnownObjOperand(knot->getOrCreateIndexAt(&appendixAddress)));
```
If `knot` is null, we still need to push an operand (unknown operand) to maintain the right stack shape.

liqunl commented:
@nbhuiyan Thanks for all the updates, it's a huge change. My remaining request is to make `appendixAddressFromXXXTable` and `memberNameAddressFromXXXTable` return address instead of the actual content, which is java object reference. And the places that use them need to be updated as well.

nbhuiyan commented:
@liqunl Thanks for the thorough review and the feedback! I have addressed your comments.

liqunl commented on runtime/compiler/optimizer/InterpreterEmulator.cpp:
```diff
@@ -621,13 +651,59 @@ InterpreterEmulator::visitInvokedynamic()
       TR_CallSite *callsite = new (comp()->trHeapMemory()) TR_J9MethodHandleCallSite(_calltarget->_calleeMethod, callNodeTreeTop,   parent,
                                                                         callNode, interfaceMethod, resolvedMethod->classOfMethod(),
-                                                                        -1, cpIndex, resolvedMethod,
+                                                                        -1, callSiteIndex, resolvedMethod,
```
We should pass -1 as `cpIndex` since `callSiteIndex` is not `cpIndex`

liqunl commented:
Except one minor issue, the changes look good to me.

nbhuiyan commented:
@andrewcraik having addressed Liqun's final comment, I think this PR is ready for your review.

liqunl commented:
@andrewcraik Should we add the single-slot change in a separate PR? Some of the code in current implementation will have to change for the single-slot design discussed in https://github.com/eclipse/openj9/pull/10762#issuecomment-706185347.Merging this PR first can leave current implementation in git history and we can easily try it should there a performance issue with single-slot design. And we have to wait for VM's implementation first to test the single-slot change. Maybe we can merge this one first?

andrewcraik commented:
Perhaps we could just add a second commit to this PR to change from two slot to one slot that way we have the history but can just put in the final version rather than merging something that isn't going to be right when the VM part merges?

nbhuiyan commented:
I prefer adding a follow-up commit adapting this to using the single slot design. Since we need the VM implementation to test the change, I will change this PR to Draft for now.

nbhuiyan commented:
@liqunl @0xdaryl This PR is ready for review.

liqunl commented on runtime/compiler/optimizer/InterpreterEmulator.cpp:
```diff
@@ -594,7 +597,7 @@ InterpreterEmulator::prepareToFindAndCreateCallsites(TR::Block **blocks, flags8_
 InterpreterEmulator::visitInvokedynamic()
    {
-   int32_t cpIndex = next2Bytes();
+
```
We should do nothing for AOT compile for now.

liqunl commented on runtime/compiler/optimizer/InterpreterEmulator.cpp:
```diff
@@ -621,13 +656,64 @@ InterpreterEmulator::visitInvokedynamic()
+#if defined(J9VM_OPT_OPENJDK_METHODHANDLE)
+void
+InterpreterEmulator::visitInvokehandle()
+   {
```
We should do nothing for AOT compile for now

nbhuiyan commented:
@liqunl @0xdaryl Requesting review

0xdaryl commented on runtime/compiler/codegen/J9CodeGenerator.cpp:
```diff
@@ -1075,8 +1075,19 @@ J9::CodeGenerator::lowerTreeIfNeeded(
        (node->getSymbol()->castToMethodSymbol()->getRecognizedMethod() == TR::java_lang_invoke_MethodHandle_invokeBasic))
       {
       TR::SymbolReference *vmThreadTempSlotSymRef = self()->comp()->getSymRefTab()->findOrCreateVMThreadTempSlotFieldSymbolRef();
```
It would be instructive to have a brief comment here describing what this lowering is doing rather than having to grok the code.  Presumably, this is a meta-argument for invokeBasic?

0xdaryl commented on runtime/compiler/compile/J9SymbolReferenceTable.cpp:
```diff
@@ -533,6 +557,45 @@ J9::SymbolReferenceTable::findOrCreateMethodTypeTableEntrySymbol(TR::ResolvedMet
+
+   TR::SymbolReference *newRef = findOrCreateSymRefWithKnownObject(originalSymRef, arrayElementKnotIndex);
+   return newRef;
+   }
```
The logic in the invokeHandle and invokeDynamic cases is 95% identical.  Can you think of a way of combining the internals into a shared function without impacting understanding?    I'm just thinking of the code footprint here...

0xdaryl commented on runtime/compiler/env/VMJ9.h:
```diff
@@ -794,6 +794,41 @@ class TR_J9VMBase : public TR_FrontEnd
+    * \param romMethodSignature the ROM Method signature to be processed
+    * \return char * the signature for linkToStatic
+    */
+   char *                getSignatureForLinkToStaticForInvokeHandle(TR::Compilation* comp, J9UTF8* romMethodSignature);
```
Just curious why there is so much whitespace preceding the function name.  It doesn't match the style of the surrounding code.

0xdaryl commented on runtime/compiler/env/VMJ9.h:
```diff
@@ -794,6 +794,41 @@ class TR_J9VMBase : public TR_FrontEnd
+    * \param romMethodSignature the ROM Method signature to be processed
+    * \return char * the signature for linkToStatic
+    */
+   char *                getSignatureForLinkToStaticForInvokeDynamic(TR::Compilation* comp, J9UTF8* romMethodSignature);
```
Just curious why there is so much whitespace preceding the function name.  It doesn't match the style of the surrounding code.

nbhuiyan commented:
@0xdaryl I have addressed your comments.

0xdaryl commented:
Jenkins test sanity all jdknext

0xdaryl commented:
@liqunl : can you formally approve if you're ok with these changes please?

liqunl commented on runtime/compiler/compile/J9SymbolReferenceTable.cpp:
```diff
@@ -533,6 +557,23 @@ J9::SymbolReferenceTable::findOrCreateMethodTypeTableEntrySymbol(TR::ResolvedMet
+#if defined(J9VM_OPT_OPENJDK_METHODHANDLE)
+TR::SymbolReference *
+J9::SymbolReferenceTable::refineInvokeCacheElementSymRefWithKnownObjectIndex(TR::ResolvedMethodSymbol * owningMethodSymbol,  TR::SymbolReference * originalSymRef, uintptr_t arrayElementRef)
+   {
```
Since we're taking an object ref, VM access should have been acquired by the caller. We need an assert here to ensure that

liqunl commented:
`loadFromSideTableForInvokeDynamic` and `loadFromSideTableForInvokeHandle` share some common logic, can we reduce the code duplication? Same to the inliner change `visitInvokedynamic` and `visitInvokehandle`

harryyu1994 commented:
Question regarding `genInvokeDynamic()` and `visitInvokeDynamic()`
- genInvokeDynamic looks at bytecode and generate IL using callSiteIndex
- visitInvokeDynamic (part of the inliner?) again looks at bytecode and generate callSite using callSiteIndex
To me, something seems redundant over here, why is `visitInvokeDynamic()` looking at bytecode instead of IL?

andrewcraik commented:
@harryyu1994  from memory: The `visit*` methods are part of an abstract bytecode interpreter which is used to propagate types through method handle chains and invokedynamic calls so OpenJ9 can see what things are available to inline. This is done on bytecode and not trees because generating trees for all the methods that you might want to consider during the inlining discovery phase is very expensive in terms of time and space. The genInvokeDynamic is about generating trees to represent the invoke dynamic in the walker so these methods are part of the tree generation.

So are they similar - yes in that they both model the semantics of the invoke dynamic bytecode. They, however, serve very different purposes.

0xdaryl commented:
Jenkins test sanity all jdk8,jdk11,jdk16

0xdaryl commented:
@liqunl : can you confirm that your concerns were addressed in the latest push and formally approve if so?  Thanks.

liqunl commented:
`loadFromSideTableForInvokeDynamic` and `loadFromSideTableForInvokeHandle` is probably not necessary as they're only used once. Can we put their code to `genInvokeDynamic` and `genInvokeHandle` respectively? 
Also, can we mark places that don't work with JITServer and AOT with comments? Such as the places we dereference `entryLocation`

liqunl commented on runtime/compiler/optimizer/InterpreterEmulator.cpp:
```diff
@@ -621,14 +629,72 @@ InterpreterEmulator::visitInvokedynamic()
+         {
+         TR::VMAccessCriticalSection vmAccess(fej9);
+         uintptr_t appendixElementRef = (uintptr_t) fej9->getReferenceElement(*invokeCacheArray, JSR292_invokeCacheArrayAppendixIndex); // dereferencing invokeCacheArray cannot be done in JITServer
+         push(new (trStackMemory()) KnownObjOperand(knot->getOrCreateIndex(appendixElementRef, true)));
```
We shouldn't pass `true` to `getOrCreateIndex` as appendix element is not an array

liqunl commented on runtime/compiler/optimizer/InterpreterEmulator.cpp:
```diff
@@ -594,7 +598,13 @@ InterpreterEmulator::prepareToFindAndCreateCallsites(TR::Block **blocks, flags8_
+   TR_ResolvedJ9Method * owningMethod = static_cast<TR_ResolvedJ9Method*>(_methodSymbol->getResolvedMethod());
+   int32_t callSiteIndex = next2Bytes();
+#if defined(J9VM_OPT_OPENJDK_METHODHANDLE)
+   if (owningMethod->isUnresolvedCallSiteTableEntry(callSiteIndex) || comp()->compileRelocatableCode()) return; // do nothing if unresolved or is AOT
```
We should also return for JITServer. Same for `visitInvokehandle`

nbhuiyan commented on runtime/compiler/ilgen/J9ByteCodeIlGenerator.hpp:
```diff
@@ -122,8 +122,21 @@ class TR_J9ByteCodeIlGenerator : public TR_IlGenerator, public TR_J9ByteCodeIter
+    *    parameters for adapter method call node.
+    *
+    * \param tableEntrySymRef the symref representing the invokeCacheArray
+    * \param invokeCacheArray the invokeCacheArray
```
Java object ref

liqunl commented on runtime/compiler/optimizer/InterpreterEmulator.cpp:
```diff
@@ -594,7 +598,15 @@ InterpreterEmulator::prepareToFindAndCreateCallsites(TR::Block **blocks, flags8_
+#if defined(J9VM_OPT_OPENJDK_METHODHANDLE)
+   if (owningMethod->isUnresolvedCallSiteTableEntry(callSiteIndex) ||
+         comp()->compileRelocatableCode() ||
+         comp()->isOutOfProcessCompilation()) return; // do nothing if unresolved, is AOT or JITServer compilation
```
@dmitry-ten Should this query be guarded by `J9VM_OPT_JITSERVER` macro?

dmitry-ten replied:
Right, it should be.

0xdaryl commented:
I can do a final review once @liqunl is satisfied her comments are resolved.

liqunl commented on runtime/compiler/compile/J9SymbolReferenceTable.cpp:
```diff
@@ -533,6 +556,23 @@ J9::SymbolReferenceTable::findOrCreateMethodTypeTableEntrySymbol(TR::ResolvedMet
+   TR_ASSERT(fej9->haveAccess(), "Require VM access to be acquired by caller");
+   TR::KnownObjectTable *knot = comp()->getOrCreateKnownObjectTable();
+   if (!knot) return originalSymRef;
+   TR_ResolvedJ9Method *owningMethod = static_cast<TR_ResolvedJ9Method*>(owningMethodSymbol->getResolvedMethod());
```
owningMethod and owningMethodSymbol are not used

nbhuiyan replied:
this was left over from a previous revision. fixed.

liqunl commented on runtime/compiler/env/VMJ9.cpp:
```diff
@@ -4853,6 +4853,48 @@ TR_J9VMBase::targetMethodFromMethodHandle(uintptr_t methodHandle)
+char *
+TR_J9VMBase::getSignatureForLinkToStaticForInvokeHandle(TR::Compilation* comp, J9UTF8* romMethodSignature, int32_t &signatureLength)
+   {
+   signatureLength = J9UTF8_LENGTH(romMethodSignature) + 55;
```
Can we have a comment on what 55 is for?

liqunl commented on runtime/compiler/env/VMJ9.cpp:
```diff
@@ -4853,6 +4853,48 @@ TR_J9VMBase::targetMethodFromMethodHandle(uintptr_t methodHandle)
+char *
+TR_J9VMBase::getSignatureForLinkToStaticForInvokeDynamic(TR::Compilation* comp, J9UTF8* romMethodSignature, int32_t &signatureLength)
+   {
+   signatureLength = J9UTF8_LENGTH(romMethodSignature) + 37;
```
Same here, a comment to explain 37

liqunl commented:
Just some nitpicking, the overall changes LGTM

nbhuiyan commented on runtime/compiler/ilgen/Walker.cpp:
```diff
@@ -3311,6 +3354,48 @@ TR_J9ByteCodeIlGenerator::genInvokeHandle(int32_t cpIndex)
+   // Call generated when methodType table entry is unresolved:
+   // ------------------------------------------------------
+   // ResolveCHK
+   //    aload <CallSiteTableEntry @<callSiteIndex>>
```
good catch!

0xdaryl commented:
Jenkins test sanity all jdknext

0xdaryl commented:
Jenkins test sanity all jdknext

nbhuiyan commented:
@liqunl I had to make a last-minute [change in `findOrCreateDynamicMethodSymbol`](https://github.com/eclipse/openj9/compare/b713a099888116bc96eb089c48bf4ddbd4b7a7bf..a465fc3cac2b176f4807a057449b6d648d9ee9a8) as we cannot use the `dynamicMethodSymrefsByCallSiteIndex` the same way as the old implementation given as it can result in wrong decisions about what needs to be loaded on the stack, given how we changed the code recently to only make a single check on whether the call site table entry is resolved in the entire process of finding/creating the dynamic method symbols.

liqunl commented:
The new update LGTM
